### PR TITLE
START_GATE

### PR DESCRIPTION
**Pull request by RustyBamboo**
- More thorough gate segmentation
- Uses stereo
Still have a few TODOs...

### CODE REVIEW

File: command/sub8_missions/missions/start_gate.py [hunk 1]
```diff
@@ -1,118 +1,145 @@
...
 
 fprint = text_effects.FprintFactory(
     title="START_GATE", msg_color="white").fprint
 
 
 class StartGateMission(object):
-
-    FOUND_START_GATE = False
     SPEED = 0.3
+    HOW_MANY = 20
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: I don't think this is great naming practice. Maybe `FEATURE_COUNT`?

File: command/sub8_missions/missions/start_gate.py [hunk 2]
```diff
@@ -1,118 +1,145 @@
...
 
     def __init__(self, sub):
-        self.sub_singleton = sub
+        # TODO: vision proxy
+        self._sub = sub
+        self._how_many_points = 0
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: Sorry for nitpicking but this just looks odd to me, I'd suggest `point_count`

File: command/sub8_missions/missions/start_gate.py [hunk 3]
```diff
@@ -1,118 +1,145 @@
...
+        self._center_pt_avg.z /= self._how_many_points
+
+        fprint('The normal: {} '.format(self._normal_avg))
+        fprint('The point: {} '.format(self._center_pt_avg))
 
         '''
-        while(self.FOUND_START_GATE):
-            self.start_gate_find()
-            yield self.sub_singleton.move.forward(.3).zero_roll_and_pitch().go(speed=self.SPEED)
+            Go directly in front of gate
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: You should use a single line comment here

File: command/sub8_missions/missions/start_gate.py [hunk 4]
```diff
@@ -1,118 +1,145 @@
...
+
+        fprint('Going infront of gate {}'.format(transformed_before_gate._p))
+        yield self._sub.move.set_position(transformed_before_gate._p).go(speed=self.SPEED)
+
+        fprint(
+            'Going through gate... YOLO! {}'.format(
+                transformed_after_gate._p))
+        yield self._sub.move.set_position(transformed_after_gate._p).go(speed=self.SPEED)
+
+    def search_patten(self):
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: How come you're not using this search pattern?

File: command/sub8_missions/missions/start_gate.py [hunk 5]
```diff
@@ -1,118 +1,145 @@
...
-        fprint("Alignment -- Deviation: " + str(deviation), msg_color="cyan")
-        while(np.abs(deviation[0]) > error_tolerance):
-            fprint("Moving right " + str(.1 * np.sign(deviation[0])), msg_color="yellow")
-            yield self.sub_singleton.move.right(.1 * np.sign(deviation[0])).zero_roll_and_pitch().go(speed=self.SPEED)
-        while(np.abs(deviation[1]) > error_tolerance):
-            fprint("Moving up " + str(.1 * np.sign(deviation[1])), msg_color="yellow")
-            yield self.sub_singleton.move.up(.1 * np.sign(deviation[1])).zero_roll_and_pitch().go(speed=self.SPEED)
-        yield fprint("Finished Alignment -- Deviation: " + str(deviation), msg_color="cyan")
+        '''
+            Go through the gate
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: Same as above

File: command/sub8_missions/missions/start_gate.py [hunk 6]
```diff
@@ -1,118 +1,145 @@
...
+
+        transform = yield self._tf_listener.get_transform(
+            '/map',
+            '/front_stereo',
+            self._sub.nh.get_time()
+        )
+        tft = tf.Transform.from_Pose_message(pose)
+        transformed_after_gate = transform * tft
+
+        fprint('Going infront of gate {}'.format(transformed_before_gate._p))
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: Missing a space here

File: command/sub8_missions/missions/start_gate.py [hunk 7]
```diff
@@ -1,118 +1,145 @@
...
+        pose.position = Point(self._center_pt_avg.x + self._normal_avg.x * self._start_away_from_gate,
+                              self._center_pt_avg.y + self._normal_avg.y * self._start_away_from_gate,
+                              self._center_pt_avg.z + self._normal_avg.z * self._start_away_from_gate)
+        transform = yield self._tf_listener.get_transform(
+            '/map',
+            '/front_stereo',
+            self._sub.nh.get_time()
+        )
+        tft = tf.Transform.from_Pose_message(pose)
+        transformed_before_gate = transform * tft
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: Don't say transformed, specify the frame.

File: command/sub8_missions/missions/start_gate.py [hunk 8]
```diff
@@ -1,118 +1,145 @@
...
+            normal.z = -normal.z
+        self._normal_avg.x += normal.x
+        self._normal_avg.y += normal.y
+        self._normal_avg.z += normal.z
+
+        if(self._how_many_normals > self.HOW_MANY):
+            self._found_normal = True
 
     @txros.util.cancellableInlineCallbacks
     def run_mission(self):
```
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: This `run_mission` function is pretty long, you should refactor into smaller chunks.

File: perception/sub8_perception/include/sub8_perception/start_gate.hpp [hunk 9]
```diff
@@ -1,67 +1,105 @@
...
-  double getGateDistance();
-
-  ros::NodeHandle nh_;
-  bool running_;
-
+  mil_vision::ImageWithCameraInfo left_most_recent, right_most_recent;
+
+private:
+  // Combinations of k elements from a set of size n (indexes)
+  void combinations(uint8_t n, uint8_t k, std::vector<std::vector<uint8_t>> &idx_array);
```
Comment (DSsoto) [inline perception/sub8_perception/include/sub8_perception/start_gate.hpp]: This really should go in mil_tools, not here. Also don't forget to attribute things you didn't code yourself.

File: perception/sub8_perception/include/sub8_perception/start_gate.hpp [hunk 10]
```diff
@@ -1,67 +1,105 @@
...
+                                   cv::Matx34d right_cam_mat, cv::Mat &current_left_right,
+                                   cv::Mat &current_image_right);
+  void visualize_3d_points_rviz(const std::vector<Eigen::Vector3d> &feature_pts_3d,
+                                const std::vector<Eigen::Vector3d> &proj_pts);
+
+  ros::NodeHandle nh;
+  image_transport::CameraSubscriber left_image_sub_, right_image_sub_;
   image_transport::ImageTransport image_transport_;
-  image_transport::CameraSubscriber image_sub_;
+  image_transport::Publisher debug_image_pub_left_;
```
Comment (DSsoto) [inline perception/sub8_perception/include/sub8_perception/start_gate.hpp]: Consider concatenating multiple debug images into one for easier debugging.

File: perception/sub8_perception/include/sub8_perception/start_gate.hpp [hunk 11]
```diff
@@ -1,67 +1,105 @@
...
-  bool requestStartGateEnable(std_srvs::SetBool::Request &req, std_srvs::SetBool::Response &resp);
-  bool requestStartGateDistance(sub8_msgs::BMatrix::Request &req, sub8_msgs::BMatrix::Response &resp);
-  void findGate(const sensor_msgs::ImageConstPtr &image_msg);
-
-  double getGateDistance();
-
-  ros::NodeHandle nh_;
-  bool running_;
-
+  mil_vision::ImageWithCameraInfo left_most_recent, right_most_recent;
```
Comment (DSsoto) [inline perception/sub8_perception/include/sub8_perception/start_gate.hpp]: Did you try using ROSCameraStream? it makes a lot of things more convenient, adds buffer functionality. Heres a [simple demo](https://github.com/uf-mil/mil_common/blob/70af0ae7832b2863a4d6c7251fd20599fa6ace84/perception/mil_vision/src/camera_stream_demo.cpp).

**Review by DSsoto — COMMENTED** [review]
A couple of comments from the first half of the PR since this one is pretty long.

Be sure to think about other people debugging your code at competition between runs at competition. You wouldn't want to be debugging code where all the poses are just called 'pose' or transforms 'tft'.

We should really have a class for an arbitrary stereo perception node in c++ that everyone could inherit from, there's (through no fault of yours) a lot of repeated code here. Oh well, I guess we have to mind our priorities.

**Comment by RustyBamboo** [comment]
Added combinations function in uf-mil/mil_common#74

File: command/sub8_missions/missions/start_gate.py [hunk 12]
```diff
@@ -1,118 +1,65 @@
...
-        while(np.abs(deviation[0]) > error_tolerance):
-            fprint("Moving right " + str(.1 * np.sign(deviation[0])), msg_color="yellow")
-            yield self.sub_singleton.move.right(.1 * np.sign(deviation[0])).zero_roll_and_pitch().go(speed=self.SPEED)
-        while(np.abs(deviation[1]) > error_tolerance):
-            fprint("Moving up " + str(.1 * np.sign(deviation[1])), msg_color="yellow")
-            yield self.sub_singleton.move.up(.1 * np.sign(deviation[1])).zero_roll_and_pitch().go(speed=self.SPEED)
-        yield fprint("Finished Alignment -- Deviation: " + str(deviation), msg_color="cyan")
+    title="START_GATE", msg_color="cyan").fprint
+
+FACTOR_DISTANCE_BEFORE = 1.5
```
Comment (kev-the-dev) [inline command/sub8_missions/missions/start_gate.py]: Add comments describing what this does, what units it is in

File: command/sub8_missions/missions/start_gate.py [hunk 13]
```diff
@@ -1,118 +1,65 @@
...
+    # Get the normal vector, which is assumed to be the [1,0,0] unit vector
+    normal = tf.transformations.quaternion_matrix(
+        orientation).dot(np.array([1, 0, 0, 0]))[0:3]
+    fprint('Computed normal vector: {}'.format(normal))
+
+    # go in front of gate
+    point_before = np.array(
+        [position[0], position[1], position[2]]) + FACTOR_DISTANCE_BEFORE * normal
+    fprint('Moving infront of gate {}'.format(point_before))
+    yield sub.move.set_position(point_before).zero_roll_and_pitch().go(speed=SPEED)
```
Comment (kev-the-dev) [inline command/sub8_missions/missions/start_gate.py]: You should add ```.look_at(position)``` so it faces through the gate during the move

File: command/sub8_missions/missions/start_gate.py [hunk 14]
```diff
@@ -1,118 +1,65 @@
...
+    fprint('Gate\'s position in map is: {}'.format(position))
+    fprint('Gate\'s orientation in map is: {}'.format(orientation))
+
+    # Get the normal vector, which is assumed to be the [1,0,0] unit vector
+    normal = tf.transformations.quaternion_matrix(
+        orientation).dot(np.array([1, 0, 0, 0]))[0:3]
+    fprint('Computed normal vector: {}'.format(normal))
+
+    # go in front of gate
+    point_before = np.array(
```
Comment (kev-the-dev) [inline command/sub8_missions/missions/start_gate.py]: Isn't position already a numpy array?

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp [hunk 15]
```diff
@@ -1,216 +1,631 @@
...
+  debug_image_pub_right_ = image_transport_.advertise("right", 1, true);
+  debug_image_pub_canny_ = image_transport_.advertise("canny", 1, true);
+
+  canny_low_ = nh.param<int>("canny_low_", 100);
+  canny_ratio_ = nh.param<int>("canny_ratio_", 3.0);
+  blur_size_ = nh.param<int>("blur_size_", 1);
+  dilate_amount_ = nh.param<int>("dilate_amount_", 3);
+
+  // process 10 images per second
+  dt_ = 1 / 10;
```
Comment (kev-the-dev) [inline perception/sub8_perception/src/sub8_perception/start_gate.cpp]: Are you sure this isn't 0 because in integer rounding?

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
Good work on the refactoring. Got a few changes. 

Also, this is such a big program for this perception task, it should definitely have at least one full rostest where an image or a bag is played and the 2d segmentation is verified and 3d pose is checked for sanity. This way we can make sure 700 line diffs don't break things. Checkout the [test for path marker](https://github.com/uf-mil/SubjuGator/blob/master/perception/sub8_perception/test/test_path_marker.py) for reference

File: perception/sub8_perception/nodes/torpedo_board.cpp [hunk 16]
```diff
@@ -1460,7 +1460,7 @@ bool TorpedoBoardReprojectionCost::operator()(const T *const x, const T *const y
 // Helper Functions ///////////////////////////////////////////////////////////////////////////////
 ///////////////////////////////////////////////////////////////////////////////////////////////////
 
-void combinations(uint8_t n, uint8_t k, vector<vector<uint8_t> > &idx_array)
+void mil_tools::combinations(uint8_t n, uint8_t k, vector<vector<uint8_t> > &idx_array)
```
Comment (kev-the-dev) [inline perception/sub8_perception/nodes/torpedo_board.cpp]: You should remove the function definitions for these two from torpedo board now that they're in mil common.

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp (hunk 15 above)
Comment (RustyBamboo) [reply perception/sub8_perception/src/sub8_perception/start_gate.cpp]: Yup, fixed.

**Comment by RustyBamboo** [comment]
Added a base class for stereo

File: perception/sub8_perception/src/sub8_vision_lib/stereo_base.cpp [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+std::unique_ptr<std::vector<Eigen::Vector3d>> StereoBase::get_3d_feature_points(int max_z)
+{
+  std::vector<cv::Point> features_l, features_r;
+  features_l = get_2d_feature_points((*left_cam_stream_)[0]->image());
+  features_r = get_2d_feature_points((*right_cam_stream_)[0]->image());
+
+  std::vector<int> correspondence_pair_idxs =
+      shortest_pair_stereo_matching(features_l, features_r, left_cam_stream_->rows() * 0.02);
+
+  if (std::count(correspondence_pair_idxs.begin(), correspondence_pair_idxs.end(), -1) != 0)
```
Comment (kev-the-dev) [inline perception/sub8_perception/src/sub8_vision_lib/stereo_base.cpp]: ```correspondence_pair_idxs.size()```?

Comment (RustyBamboo) [reply perception/sub8_perception/src/sub8_vision_lib/stereo_base.cpp]: It's checking if there are any -1 in the vector

**Review by kev-the-dev — APPROVED** [review]

**Pull request merged**